tests/basic/rwlock.rs

```diff
@@ -220,3 +221,286 @@ fn rwlock_into_inner() {
+                let lock = Arc::clone(&lock);
+                thread::spawn(move || {
+                    for n in 1..3 {
+                        match lock.try_write() {
+                            Ok(mut guard) => {
+                                *guard += n;
+                            }
+                            Err(TryLockError::WouldBlock) => (),
+                            Err(_) => panic!("unexpected TryLockError"),
+                        };
+                    }
+                })
+            };
+
+            match lock.try_read() {
+                Ok(guard) => {
+                    observed_values_clone.lock().unwrap().insert(*guard);
+                }
+                Err(TryLockError::WouldBlock) => (),
+                Err(_) => panic!("unexpected TryLockError"),
+            }
+
+            write_thread.join().unwrap();
+            try_write_thread.join().unwrap();
+        },
+        None,
+    );
+
+    let observed_values = Arc::try_unwrap(observed_values).unwrap().into_inner().unwrap();
+    assert_eq!(observed_values, HashSet::from([0, 1, 2, 3, 4]));
+}
+
+/// This behavior is _sometimes_ allowed in `std`, but advised against, as it can lead to deadlocks
+/// on some platforms if the second `read` races with another thread's `write`. We conservatively
+/// rule it out in all cases to better detect potential deadlocks.
+#[test]
+#[should_panic(expected = "tried to acquire a RwLock it already holds")]
+fn double_read() {
+    check_dfs(
+        || {
+            let rwlock = RwLock::new(());
+            let _guard_1 = rwlock.read().unwrap();
+            let _guard_2 = rwlock.read();
+        },
+        None,
+    )
+}
+
+#[test]
+#[should_panic(expected = "tried to acquire a RwLock it already holds")]
+fn double_write() {
+    check_dfs(
+        || {
+            let rwlock = RwLock::new(());
+            let _guard_1 = rwlock.write().unwrap();
+            let _guard_2 = rwlock.write();
+        },
+        None,
+    )
+}
+
+#[test]
+#[should_panic(expected = "tried to acquire a RwLock it already holds")]
+fn read_upgrade() {
+    check_dfs(
+        || {
+            let rwlock = RwLock::new(());
+            let _guard_1 = rwlock.read().unwrap();
+            let _guard_2 = rwlock.write();
+        },
+        None,
+    )
+}
+
+#[test]
+#[should_panic(expected = "tried to acquire a RwLock it already holds")]
+fn write_downgrade() {
+    check_dfs(
+        || {
+            let rwlock = RwLock::new(());
+            let _guard_1 = rwlock.write().unwrap();
+            let _guard_2 = rwlock.read();
+        },
+        None,
+    )
+}
+
+/// This behavior isn't consistent with `std`, which seems to suggest that `try_read` succeeds if
+/// the current thread already holds a read lock. We assume it always fails so that we can more
+/// easily diagnose potential deadlocks, especially with async tasks that might migrate across
+/// threads in real implementations.
+#[test]
+fn double_try_read() {
```

Comment:
There's a variant of this where a thread does a `read()` and then a `try_read()`

Reply:
I figured this would be the same as if the first `try_read` succeeds (assuming that a successful `try_read` and a `read` are equivalent), but may as well add the explicit test anyway.